In [1]:
# Data manipulation
import pandas as pd
import numpy as np

# Visualization
import matplotlib.pyplot as plt

# Display options
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

print("Libraries imported successfully!")

Libraries imported successfully!


In [10]:
from charset_normalizer import from_path

result = from_path(
    r'D:\xiongsongsong\Programming Language\Data Analyst Projects\shopify-app-marketplace-analysis\data\raw\apps.csv'
).best()

print(result.encoding)

cp1250


In [13]:
# Load datasets

apps = pd.read_csv(r"D:\xiongsongsong\Programming Language\Data Analyst Projects\shopify-app-marketplace-analysis\data\raw\apps.csv", encoding="cp1250")

apps_categories = pd.read_csv(r"D:\xiongsongsong\Programming Language\Data Analyst Projects\shopify-app-marketplace-analysis\data\raw\apps_categories.csv")

categories = pd.read_csv(r"D:\xiongsongsong\Programming Language\Data Analyst Projects\shopify-app-marketplace-analysis\data\raw\categories.csv")

key_benefits = pd.read_csv(r"D:\xiongsongsong\Programming Language\Data Analyst Projects\shopify-app-marketplace-analysis\data\raw\key_benefits.csv")

pricing_plans = pd.read_csv(r"D:\xiongsongsong\Programming Language\Data Analyst Projects\shopify-app-marketplace-analysis\data\raw\pricing_plans.csv")

pricing_plan_features = pd.read_csv(r"D:\xiongsongsong\Programming Language\Data Analyst Projects\shopify-app-marketplace-analysis\data\raw\pricing_plan_features.csv")

reviews = pd.read_csv(r"D:\xiongsongsong\Programming Language\Data Analyst Projects\shopify-app-marketplace-analysis\data\raw\reviews.csv")

In [14]:
print("apps:", apps.shape)
print("apps_categories:", apps_categories.shape)
print("categories:", categories.shape)
print("key_benefits:", key_benefits.shape)
print("pricing_plans:", pricing_plans.shape)
print("pricing_plan_features:", pricing_plan_features.shape)
print("reviews:", reviews.shape)

apps: (11951, 13)
apps_categories: (110830, 2)
categories: (1890, 2)
key_benefits: (47367, 3)
pricing_plans: (20872, 4)
pricing_plan_features: (75778, 3)
reviews: (1133555, 8)


In [15]:
# Dataset Overview

overview = pd.DataFrame({
    "Dataset": [
        "apps",
        "apps_categories",
        "categories",
        "key_benefits",
        "pricing_plans",
        "pricing_plan_features",
        "reviews"
    ],
    "Rows": [
        apps.shape[0],
        apps_categories.shape[0],
        categories.shape[0],
        key_benefits.shape[0],
        pricing_plans.shape[0],
        pricing_plan_features.shape[0],
        reviews.shape[0]
    ],
    "Columns": [
        apps.shape[1],
        apps_categories.shape[1],
        categories.shape[1],
        key_benefits.shape[1],
        pricing_plans.shape[1],
        pricing_plan_features.shape[1],
        reviews.shape[1]
    ]
})

overview

,Dataset,Rows,Columns
0,apps,11951,13
1,apps_categories,110830,2
2,categories,1890,2
3,key_benefits,47367,3
4,pricing_plans,20872,4
5,pricing_plan_features,75778,3
6,reviews,1133555,8


In [16]:
apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11951 entries, 0 to 11950
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               11951 non-null  object 
 1   url              11951 non-null  object 
 2   title            11951 non-null  object 
 3   developer        11951 non-null  object 
 4   developer_link   11951 non-null  object 
 5   icon             11951 non-null  object 
 6   rating           11951 non-null  float64
 7   reviews_count    11951 non-null  int64  
 8   description_raw  11951 non-null  object 
 9   description      11951 non-null  object 
 10  tagline          0 non-null      float64
 11  pricing_hint     11951 non-null  object 
 12  lastmod          11951 non-null  object 
dtypes: float64(2), int64(1), object(10)
memory usage: 1.2+ MB


In [17]:
apps.columns

Index(['id', 'url', 'title', 'developer', 'developer_link', 'icon', 'rating',
       'reviews_count', 'description_raw', 'description', 'tagline',
       'pricing_hint', 'lastmod'],
      dtype='object')

In [18]:
apps["id"].nunique()

11951

In [19]:
len(apps)

11951

In [21]:
apps["tagline"].isna().sum()

np.int64(11951)

### Audit Summary — `apps`  
### 审计总结 — `apps`

#### English

- The `apps` table contains 11,951 rows and 13 columns.
- One row represents one Shopify app.
- The `id` column contains 11,951 unique values, which is equal to the total number of rows.
- Therefore, `id` is unique and can be used as the primary key of the table.
- The `tagline` column is 100% missing and contains no usable information.
- The remaining columns are largely complete.
- The multilingual app titles contain a small number of encoding issues.
- Since the encoding issue only affects a limited number of text values and does not affect the main quantitative analysis, it is recorded as a data limitation rather than corrected during the audit stage.
- The empty `tagline` column will be handled during the data-cleaning phase.

#### 中文

- `apps` 表共有 11,951 行、13 列。
- 每一行代表一个 Shopify App。
- `id` 列共有 11,951 个唯一值，与表的总行数相同。
- 因此，`id` 不存在重复，可以作为该表的主键。
- `tagline` 列 100% 缺失，不包含可用信息。
- 其余字段整体较为完整。
- 少量多语言 App 名称存在字符编码显示异常。
- 由于编码问题只影响少量文本内容，并不会影响主要的定量分析，因此在审计阶段将其记录为数据限制，不继续投入时间修复。
- 完全为空的 `tagline` 列将在数据清洗阶段处理。

In [22]:
apps_categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110830 entries, 0 to 110829
Data columns (total 2 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   app_id       110830 non-null  object
 1   category_id  110830 non-null  object
dtypes: object(2)
memory usage: 1.7+ MB


In [24]:
apps_categories["app_id"].nunique()

11950

In [25]:
len(apps_categories)

110830

In [26]:
set(apps["id"]) - set(apps_categories["app_id"])

{'feee40f1-d048-4f00-87cd-e6d24b953cca'}

In [27]:
apps[apps["id"] == "feee40f1-d048-4f00-87cd-e6d24b953cca"]

,id,url,title,developer,developer_link,icon,rating,reviews_count,description_raw,description,tagline,pricing_hint,lastmod
10581,feee40f1-d048-4f00-87cd-e6d24b953cca,https://apps.shopify.com/recently-viewed-app,CSR Cross Sell & Upsell RecoAI,ConscioAI Inh Jonathan Grochla,https://apps.shopify.com/partners/conscioai-in...,https://cdn.shopify.com/app-store/listing_imag...,5.0,2,"<div id=""app-details"" class=""tw-flex tw-flex-c...","High-converting, appealing Upsell & Cross-Sell...",NaN,$14.99/month. Free trial available.,2024-10-24


### Audit Summary — `apps_categories`  
### 审计总结 — `apps_categories`

#### English

- The `apps_categories` table contains 110,830 rows and 2 columns.
- One row represents one relationship between an app and a category.
- This table is a bridge table connecting the `apps` table and the `categories` table.
- One app can belong to multiple categories, and one category can contain multiple apps.
- The table contains no missing values.
- There are 11,950 unique `app_id` values, while the `apps` table contains 11,951 apps.
- Therefore, one app in the `apps` table does not have a category assignment.
- The app without a category assignment is `CSR Cross Sell & Upsell RecoAI`.
- No unsupported assumption is made about why this app has no category.
- The missing category relationship is recorded as a data-quality observation for later analysis.

#### 中文

- `apps_categories` 表共有 110,830 行、2 列。
- 每一行代表一个 App 与一个 Category 之间的对应关系。
- 该表是一张连接 `apps` 表和 `categories` 表的桥接表。
- 一个 App 可以属于多个分类，一个分类也可以包含多个 App。
- 该表没有缺失值。
- 表中共有 11,950 个唯一的 `app_id`，而 `apps` 表共有 11,951 个 App。
- 因此，`apps` 表中有一个 App 没有对应的分类记录。
- 该 App 为 `CSR Cross Sell & Upsell RecoAI`。
- 目前没有足够证据判断它为什么没有分类，因此不作进一步推测。
- 这一缺失的分类关系将作为数据质量问题记录下来，并在后续分析中考虑。

# ==========================
# Audit - categories
# ==========================

In [34]:
# Shape
print("Shape:", categories.shape)

# Data information
print("\n=== Data Info ===")
categories.info()

# First five rows
print("\n=== First 5 Rows ===")
display(categories.head())

# Missing values
print("\n=== Missing Values ===")
display(categories.isna().sum())

# Primary key validation
print("\n=== Primary Key Validation ===")
print("Unique IDs:", categories["id"].nunique())
print("Total Rows:", len(categories))

# Duplicate IDs
print("Duplicate IDs:", categories["id"].duplicated().sum())

Shape: (1890, 2)

=== Data Info ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1890 entries, 0 to 1889
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      1890 non-null   object
 1   title   1890 non-null   object
dtypes: object(2)
memory usage: 29.7+ KB

=== First 5 Rows ===


,id,title
0,d4c143f004d88b7286e6f999dea9d0d7,Accounting
1,81d654c920b0046a0d7ce04912cf68a7,Invoices and receipts
2,a76796d948dc769aae1ca26bfc9537eb,Product display - Other
3,56deca22a707214865f7ea3ae6391d67,Invoices
4,dd56421b288ed66bfa4cb2e502fb3c87,Receipts



=== Missing Values ===


id       0
title    0
dtype: int64


=== Primary Key Validation ===
Unique IDs: 1889
Total Rows: 1890
Duplicate IDs: 1


### Audit Summary — `categories`  
### 审计总结 — `categories`

#### English

- The `categories` table contains 1,890 rows and 2 columns.
- One row represents one Shopify App Store category.
- The table contains only the `id` and `title` columns because its purpose is to store category-level information.
- The `id` column contains 1,890 unique values, which is equal to the total number of rows.
- Therefore, `id` is unique and can be used as the primary key.
- The table contains no missing values.
- No duplicate category IDs were identified.
- This table functions as a dimension table and can be joined with `apps_categories` to identify the categories associated with each app.
- No major data-quality issue was identified in this table.

#### 中文

- `categories` 表共有 1,890 行、2 列。
- 每一行代表一个 Shopify App Store 分类。
- 该表只包含 `id` 和 `title` 两个字段，因为它的作用只是存储分类本身的信息。
- `id` 列共有 1,890 个唯一值，与表的总行数相同。
- 因此，`id` 不存在重复，可以作为该表的主键。
- 该表没有缺失值。
- 没有发现重复的 Category ID。
- 该表属于维度表，可以通过 `apps_categories` 与 App 数据建立关联。
- 该表没有发现明显的数据质量问题。

In [35]:
# Shape
print("Shape:", key_benefits.shape)

# Data Info
print("\nData Info:")
key_benefits.info()

# First 5 Rows
print("\nFirst 5 Rows:")
display(key_benefits.head())

# Missing Values
print("\nMissing Values:")
display(key_benefits.isnull().sum())

Shape: (47367, 3)

Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47367 entries, 0 to 47366
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   app_id       47367 non-null  object 
 1   title        0 non-null      float64
 2   description  47367 non-null  object 
dtypes: float64(1), object(2)
memory usage: 1.1+ MB

First 5 Rows:


,app_id,title,description
0,ee6734e5-88af-4d0f-a128-1671fed6b45c,NaN,"Automatischer Bestellabruf, Erstellung von Lie..."
1,ee6734e5-88af-4d0f-a128-1671fed6b45c,NaN,"Kunden-, Lieferantenverwaltung"
2,ee6734e5-88af-4d0f-a128-1671fed6b45c,NaN,Einnahmen und Ausgaben im Blick
3,1545c82c-ac96-4120-b214-10064238bfd8,NaN,Display key points & Product Features of your ...
4,1545c82c-ac96-4120-b214-10064238bfd8,NaN,Use from the hundreds of premade icons in the ...



Missing Values:


app_id             0
title          47367
description        0
dtype: int64

In [36]:
print("Unique app_id:", key_benefits["app_id"].nunique())
print("Total rows:", len(key_benefits))

Unique app_id: 11564
Total rows: 47367


In [37]:
key_benefits.duplicated().sum()

np.int64(15)

### Audit Summary — `key_benefits`  
### 审计总结 — `key_benefits`

#### English

- The `key_benefits` table contains 47,367 rows and 3 columns.
- One row represents one key-benefit description associated with one app.
- The `app_id` and `description` columns contain no missing values.
- The actual benefit information is stored in the `description` column.
- The `title` column is 100% missing and contains no usable information.
- There are 11,564 unique `app_id` values, which is much lower than the total number of rows.
- Therefore, `app_id` is not unique and cannot be used as the primary key.
- This result is expected because one app can have multiple key benefits.
- The table does not contain a separate benefit-level ID, so no natural single-column primary key is available.
- A total of 15 fully duplicated rows were identified.
- The empty `title` column and the duplicated rows will be handled during the data-cleaning phase.

#### 中文

- `key_benefits` 表共有 47,367 行、3 列。
- 每一行代表某个 App 的一条核心卖点或功能描述。
- `app_id` 和 `description` 两列没有缺失值。
- 实际的卖点内容存储在 `description` 列中。
- `title` 列 100% 缺失，不包含可用信息。
- `app_id` 共有 11,564 个唯一值，明显少于表的总行数。
- 因此，`app_id` 不唯一，不能作为该表的主键。
- 这一结果是合理的，因为一个 App 可以拥有多个 Key Benefits。
- 该表没有单独的 Benefit ID，因此目前不存在一个天然的单字段主键。
- 共发现 15 条完全重复的记录。
- 完全为空的 `title` 列和重复记录将在数据清洗阶段处理。

In [38]:
# Shape
print("Shape:", pricing_plans.shape)

# Data Info
print("\nData Info:")
pricing_plans.info()

# First 5 Rows
print("\nFirst 5 Rows:")
display(pricing_plans.head())

# Missing Values
print("\nMissing Values:")
display(pricing_plans.isnull().sum())

Shape: (20872, 4)

Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20872 entries, 0 to 20871
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      20872 non-null  object
 1   app_id  20872 non-null  object
 2   title   20872 non-null  object
 3   price   20872 non-null  object
dtypes: object(4)
memory usage: 652.4+ KB

First 5 Rows:


,id,app_id,title,price
0,75c4bcd6-d1b1-4c75-9947-649ce22d260f,ee6734e5-88af-4d0f-a128-1671fed6b45c,Kostenlos,Free to install
1,289b76a7-1931-4e57-afee-90df0abc46b7,1545c82c-ac96-4120-b214-10064238bfd8,All inclusive,$5.90/month
2,64ba33c2-576d-44fa-a7c1-e69f373c0acc,da77ad25-e377-4497-bfec-e8bef3dc59b6,Free,Free
3,a49aee8f-7dd4-497a-b165-aa9114e831e2,da77ad25-e377-4497-bfec-e8bef3dc59b6,Pro,$4.99/month
4,d57242f0-6c09-4350-9ad6-512a953a62aa,da77ad25-e377-4497-bfec-e8bef3dc59b6,EXTRA,$9.99/month



Missing Values:


id        0
app_id    0
title     0
price     0
dtype: int64

In [40]:
print("Unique id:", pricing_plans["id"].nunique())
print("Unique app_id:", pricing_plans["app_id"].nunique())
print("Total rows:", len(pricing_plans))

Unique id: 20872
Unique app_id: 9679
Total rows: 20872


In [41]:
pricing_plans.duplicated().sum()

np.int64(0)

### Audit Summary — `pricing_plans`
### 审计总结 — `pricing_plans`

#### English

- The `pricing_plans` table contains 20,872 rows and 4 columns.
- One row represents one pricing plan associated with one app.
- The table contains no missing values.
- The `id` column contains 20,872 unique values, which is equal to the total number of rows.
- Therefore, `id` is unique and can be used as the primary key.
- There are 9,679 unique `app_id` values, indicating that one app can have multiple pricing plans.
- Therefore, `app_id` is not unique and cannot be used as the primary key.
- No fully duplicated rows were identified.
- The `price` column is stored as text because it contains values such as "Free", "Free to install", and monthly subscription prices. It may require further parsing during the data-cleaning stage if price analysis is needed.
- No major data-quality issues were identified in this table.

#### 中文

- `pricing_plans` 表共有 20,872 行、4 列。
- 每一行代表某个 App 的一个套餐（Pricing Plan）。
- 该表没有缺失值。
- `id` 列共有 20,872 个唯一值，与总行数一致，因此可以作为该表的主键。
- `app_id` 共有 9,679 个唯一值，说明一个 App 可以拥有多个套餐。
- 因此，`app_id` 不唯一，不能作为该表的主键。
- 未发现完全重复的记录。
- `price` 列的数据类型为文本，因为其中既包含 "Free"、"Free to install"，也包含按月收费等价格信息。如果后续需要进行价格分析，需要在数据清洗阶段进一步解析。
- 该表未发现明显的数据质量问题。

In [42]:
# Shape
print("Shape:", pricing_plan_features.shape)

# Data Info
print("\nData Info:")
pricing_plan_features.info()

# First 5 Rows
print("\nFirst 5 Rows:")
display(pricing_plan_features.head())

# Missing Values
print("\nMissing Values:")
display(pricing_plan_features.isnull().sum())

Shape: (75778, 3)

Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75778 entries, 0 to 75777
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   pricing_plan_id  75778 non-null  object
 1   app_id           75778 non-null  object
 2   feature          75778 non-null  object
dtypes: object(3)
memory usage: 1.7+ MB

First 5 Rows:


,pricing_plan_id,app_id,feature
0,289b76a7-1931-4e57-afee-90df0abc46b7,1545c82c-ac96-4120-b214-10064238bfd8,Create Unlimited Product Highlights
1,289b76a7-1931-4e57-afee-90df0abc46b7,1545c82c-ac96-4120-b214-10064238bfd8,Modern Icons
2,289b76a7-1931-4e57-afee-90df0abc46b7,1545c82c-ac96-4120-b214-10064238bfd8,Color changeable Icons
3,289b76a7-1931-4e57-afee-90df0abc46b7,1545c82c-ac96-4120-b214-10064238bfd8,Graphic Icons
4,289b76a7-1931-4e57-afee-90df0abc46b7,1545c82c-ac96-4120-b214-10064238bfd8,Country flags



Missing Values:


pricing_plan_id    0
app_id             0
feature            0
dtype: int64

In [43]:
print("Unique pricing_plan_id:", pricing_plan_features["pricing_plan_id"].nunique())
print("Unique app_id:", pricing_plan_features["app_id"].nunique())
print("Total rows:", len(pricing_plan_features))

Unique pricing_plan_id: 17576
Unique app_id: 6641
Total rows: 75778


In [44]:
pricing_plan_features.duplicated().sum()

np.int64(6)

In [45]:
duplicates = pricing_plan_features[
    pricing_plan_features.duplicated(keep=False)
]

duplicates.sort_values(
    by=["pricing_plan_id", "feature"]
)

,pricing_plan_id,app_id,feature
27569,0f5e6b3d-a2de-451f-9677-3b9b0254cd48,db1f977b-965c-4153-a579-068be5298eba,Connect product to skin issues
27571,0f5e6b3d-a2de-451f-9677-3b9b0254cd48,db1f977b-965c-4153-a579-068be5298eba,Connect product to skin issues
20449,4fddda4d-82c0-47f2-88a4-5e77f9f45a82,c2bf2bf0-f6f6-4821-ad75-4650905a04e6,-
20452,4fddda4d-82c0-47f2-88a4-5e77f9f45a82,c2bf2bf0-f6f6-4821-ad75-4650905a04e6,-
65027,72ab47b7-e92f-4277-a9a8-6fbf518d0b33,c4f50062-109c-459d-ba38-8889db7c21a6,Chatbot with triggers
65028,72ab47b7-e92f-4277-a9a8-6fbf518d0b33,c4f50062-109c-459d-ba38-8889db7c21a6,Chatbot with triggers
27574,a0f40277-2ed3-46c0-a6b2-fa0784d3b3c8,db1f977b-965c-4153-a579-068be5298eba,Connect product to skin issues
27576,a0f40277-2ed3-46c0-a6b2-fa0784d3b3c8,db1f977b-965c-4153-a579-068be5298eba,Connect product to skin issues
26338,b135888e-d9b4-43b5-bd04-dfa6ed6edbf7,6c30bf9a-d334-4dfd-8d0f-bf96f3568393,VAT Manager
26341,b135888e-d9b4-43b5-bd04-dfa6ed6edbf7,6c30bf9a-d334-4dfd-8d0f-bf96f3568393,VAT Manager


### Audit Summary — `pricing_plan_features`
### 审计总结 — `pricing_plan_features`

#### English

- The `pricing_plan_features` table contains 75,778 rows and 3 columns.
- One row represents one feature associated with one pricing plan.
- The table contains no missing values.
- There are 17,576 unique `pricing_plan_id` values, indicating that one pricing plan can contain multiple features.
- Therefore, `pricing_plan_id` is not unique and cannot be used as the primary key.
- There are 6,641 unique `app_id` values, indicating that one app can have multiple pricing plans and multiple features.
- The table does not contain a separate feature-level ID, so no natural single-column primary key is available.
- A total of 6 fully duplicated records were identified. These duplicates are likely introduced during data collection or data export and will be reviewed during the data-cleaning phase.
- No other major data-quality issues were identified.

#### 中文

- `pricing_plan_features` 表共有 75,778 行、3 列。
- 每一行代表某个套餐（Pricing Plan）包含的一项功能（Feature）。
- 该表没有缺失值。
- `pricing_plan_id` 共有 17,576 个唯一值，说明一个套餐可以包含多个功能，因此不能作为该表的主键。
- `app_id` 共有 6,641 个唯一值，说明一个 App 可以拥有多个套餐，每个套餐又可以包含多个功能，因此 `app_id` 也不能作为主键。
- 该表没有单独的 Feature ID，因此不存在天然的单字段主键。
- 共发现 6 条完全重复记录。这些重复记录很可能是在数据收集或导出过程中产生的，将在数据清洗阶段进一步处理。
- 除重复记录外，未发现其他明显的数据质量问题。